File: DataSets/DataSet-Old/LeetCodeTasks - Develop/LeetCodeTasks/84_largest-rectangle-in-histogram/solutions/python3/solution.py
class Solution:
    def largestRectangleArea(self, heights: list[int]) -> int:
        stack = []
        max_area = 0
        heights.append(0)  # Add zero height to flush out remaining bars in stack
        
        for i in range(len(heights)):
            while stack and heights[stack[-1]] > heights[i]:
                h = heights[stack.pop()]
                w = i if not stack else i - stack[-1] - 1
                max_area = max(max_area, h * w)
            stack.append(i)
        
        return max_area

File: DataSets/DataSet-Old/LeetCodeTasks - Develop/LeetCodeTasks/84_largest-rectangle-in-histogram/solutions/c/solution.c
#include <stdlib.h>

int largestRectangleArea(int* heights, int heightsSize) {
    int *stack = (int*) malloc(sizeof(int) * heightsSize);
    int top = -1;
    int maxArea = 0;
    int i = 0;
    
    while (i <= heightsSize) {
        int h = (i == heightsSize ? 0 : heights[i]);
        if (top == -1 || heights[stack[top]] <= h) {
            stack[++top] = i++;
        } else {
            int height = heights[stack[top--]];
            int width = (top == -1 ? i : i - stack[top] - 1);
            maxArea = (maxArea > height * width ? maxArea : height * width);
        }
    }
    
    free(stack);
    return maxArea;
}
